packages/webamp-modern-2/src/skin/ImageManager.ts

```diff
@@ -26,6 +39,84 @@ export default class ImageManager {
     return this._urlCache.get(filePath);
   }
 
+  // called only on RESOURCE_PHASE
+  // addBitmap(id:string, bitmap:Bitmap, filePath: string) {
+  addBitmap(bitmap:Bitmap) {
+    const id = bitmap._id;
+    const filePath = bitmap._file;
+    // this._bitmapPath[id] = filePath;
+    this._pathofBitmap[filePath] = false;
+    this._bitmaps[id] = bitmap;
+    // if (!this._imgCache.has(filePath)) {
+    //   this._imgCache.set(filePath, null);
+    // }
+  }
+
+  // called only on RESOURCE_PHASE
+  // addBitmapFont(filePath: string) {
+  //   this._bitmapAlias[filePath] = null; //?just register
+  // }
+
+  // Ensure we've loaded the image into our image loader.
+  async loadUniquePaths() {
+    // this._bitmapPath.values()
+    // const promises = [];
+    for (const filePath of Object.keys(this._pathofBitmap)) {
+      await this.getImage(filePath)
+    }
+
+    // return Promise.all(
+    //   Object.values(this._bitmaps).map( bitmap => {
+    //     return this.setBimapImg(bitmap)
+    //   })
+    // )
+  }
+  async ensureBitmapsLoaded() {
+    // this._bitmapPath.values()
+    // const promises = [];
+    // for (const [id, bitmap] of Object.entries(this._bitmaps)) {
+
+    // }
+
+    return Promise.all(
+      Object.values(this._bitmaps).map( bitmap => {
+        return this.setBimapImg(bitmap).then(
+          () => {
+            if(bitmap._img){
+              if (bitmap._width == null && bitmap._height == null) {
+                bitmap.setXmlAttr("w", String(bitmap._img.width));
+                bitmap.setXmlAttr("h", String(bitmap._img.height));
+              }
+            }
+          }
+        )
+      })
```

Comment:
I think this could be a bit cleaner:

```suggestion
      Object.values(this._bitmaps).map(async bitmap => {
          await this.setBimapImg(bitmap);
          if(bitmap._img &&bitmap._width == null && bitmap._height == null) {
              bitmap.setXmlAttr("w", String(bitmap._img.width));
              bitmap.setXmlAttr("h", String(bitmap._img.height));
          }
      })
```